XRay Guide using OpenTelemetry (#976)
micronaut-projects/micronaut-guides: guides/micronaut-microservices-distributed-tracing-xray/micronaut-microservices-distributed-tracing-xray.adoc

common:header-top.adoc[]

== Getting Started

In this guide, we are going to use https://opentelemetry.io[OpenTelemetry] to collect distributed
tracing information of a Micronaut application composed of three microservices and publish it to https://aws.amazon.com/xray/[AWS X-Ray]

____

> AWS X-Ray helps developers analyze and debug production, distributed applications, such as those built using a microservices architecture. With X-Ray, you can understand how your application and its underlying services are performing to identify and troubleshoot the root cause of performance issues and errors. X-Ray provides an end-to-end view of requests as they travel through your application, and shows a map of your application’s underlying components. You can use X-Ray to analyze both applications in development and in production, from simple three-tier applications to complex microservices applications consisting of thousands of services.

Reviewer: ```suggestion
AWS X-Ray helps developers analyze and debug distributed applications, such as those built using a microservices architecture. With X-Ray, you can understand how your application and its underlying services are performing to identify and troubleshoot the root cause of performance issues and errors. X-Ray provides an end-to-end view of requests as they travel through your application and displays a map of your application’s underlying components. You can use X-Ray to analyze both applications in development and in production, from simple three-tier applications to complex microservices applications consisting of thousands of services.
```

I know you copied this directly from the AWS site, but the first sentence doesn't make sense as written and the comma I removed does not belong there. :-)

____

You will discover how the https://micronaut-projects.github.io/micronaut-tracing/latest/guide/[Micronaut framework eases AWS X-Ray integration via OpenTelemetry].

common:requirements.adoc[]

common:completesolution.adoc[]

== Writing the application

The application contains three microservices.

> * `bookcatalogue` - It returns a list of books. It uses a domain consisting of a book name and ISBN.

Reviewer: ```suggestion
* `bookcatalogue` - Returns a list of books. It uses a domain consisting of a book name and an ISBN.
```

> * `bookinventory` - It exposes an endpoint to check whether a book has sufficient stock to fulfill an order. It uses a domain consisting of a stock level and ISBN.

Reviewer: ```suggestion
* `bookinventory` - Exposes an endpoint to check whether a book has sufficient stock to fulfill an order. It uses a domain consisting of a stock level and an ISBN.
```

> * `bookrecommendation` - It consumes previous services and exposes an endpoint recommending book names in stock.

Reviewer: ```suggestion
* `bookrecommendation` - Consumes previous services and exposes an endpoint recommending book names in stock.
```

The `bookrecommendation` service consumes endpoints exposed by the other services. The following image illustrates the application flow:

image::flow-dynamodb.svg[]

A request to `bookrecommendation` (`http://localhost:8080/books[http://localhost:8080/books^]`) triggers several requests through our microservices mesh.

common:annotationprocessors.adoc[]

== Run AWS OTel Collector locally

See https://github.com/aws-observability/aws-otel-collector/blob/main/docs/developers/docker-demo.md[aws-otel-collector/blob/main/docs/developers/docker-demo.md]

* Clone https://github.com/aws-observability/aws-otel-collector
* `vi examples/docker/config-test.yaml` and set the region
* `docker run --rm -p 4317:4317 -p 55680:55680 -p 8889:8888 -e AWS_REGION=us-east-1 -e AWS_PROFILE=default -v ~/.aws:/root/.aws -v "${PWD}/examples/docker/config-test.yaml":/otel-local-config.yaml --name awscollector public.ecr.aws/aws-observability/aws-otel-collector:latest --config otel-local-config.yaml`

== AWS Command Line Interface

Install the https://aws.amazon.com/cli/[AWS Command Line Interface] and run `aws configure`.
Set the same region you used in the previous section.

== OpenTelemetry dependencies

> === OpenTelemetry annotations

Reviewer 2: There isn't any OT annotations.

Reviewer 3: Removed 👍

To enable OpenTelemetry annotations, every service includes the next annotation processor:

dependency:micronaut-tracing-opentelemetry-annotation[groupId=io.micronaut.tracing,scope=annotationProcessor]

=== Micronaut OpenTelemetry HTTP

> To enable creating span objects on the every HTTP server request, client request, server response and client response

Reviewer: ```suggestion
To enable creation of span objects on every HTTP server request, client request, server response, and client response,
```

each service includes the next dependency:

dependency:micronaut-tracing-opentelemetry-http[groupId=io.micronaut.tracing,scope=compile]

=== OpenTelemetry Protocol Exporter

An exporter is a component in the OpenTelemetry Collector configured to send data to different systems/back-ends.

Each service adds the OpenTelemetry Protocol exporter dependency.

dependency:opentelemetry-exporter-otlp[groupId=io.opentelemetry,scope=compile]

=== OpenTelemetry Propagator

> To use AWS X-Ray propagator each service adds the following dependency:

Reviewer: ```suggestion
To use AWS X-Ray propagator, each service adds the following dependency:
```

dependency:opentelemetry-extension-aws[groupId=io.opentelemetry,scope=compile]

and the following configuration:

resource:application.yml[app=bookrecommendation,tag=otel]

The previous configuration excludes health checks from being traced.

===  Using AWS resource detector

Each service includes the following dependency:

dependency:opentelemetry-aws-xray[groupId=io.opentelemetry.contrib,scope=compile]

> They allow you to enable https://aws-otel.github.io/docs/getting-started/java-sdk/trace-manual-instr#using-the-aws-resource-detectors[AWS resource detectors for enriching traces

Reviewer: It's not clear here what "they" refers to.

```suggestion
These dependencies allow you to enable https://aws-otel.github.io/docs/getting-started/java-sdk/trace-manual-instr#using-the-aws-resource-detectors[AWS resource detectors for enriching traces
```

 with AWS infrastructure information].

=== Instrumenting the AWS SDK

`inventory` and `catalogue` include the following dependencies to instrument the https://aws-otel.github.io/docs/getting-started/java-sdk/trace-manual-instr#instrumenting-the-aws-sdk[AWS SDK]:

dependency:opentelemetry-aws-sdk-2.2[groupId=io.opentelemetry.instrumentation,scope=compile]

Both services use DynamoDB and include the following dependencies:

:dependencies:
dependency:micronaut-aws-sdk-v2[groupId=io.micronaut.aws,scope=compile]
dependency:dynamodb[groupId=software.amazon.awssdk]
:dependencies:

`micronaut-aws-sdk-v2` dependency creates a bean of type `SdkClientBuilder`. To instrument the AWS SDK,

> Micronaut OpenTelemetry registers a tracing interceptor by creating a bean creation listener for the bean of type `SdkClientBuilder`.

Reviewer: ```suggestion
Micronaut OpenTelemetry registers a tracing interceptor by creating a bean-creation listener for the bean of type `SdkClientBuilder`.
```

:exclude-for-build:gradle

To run the application, execute `./mvnw mn:run`.

:exclude-for-build:

[source,bash]
----
...
14:28:34.034 [main] INFO  io.micronaut.runtime.Micronaut - Startup completed in 499ms. Server Running: http://localhost:8081
----

Run `bookinventory` microservice:

:exclude-for-build:maven

To run the application, execute `./gradlew run`.

:exclude-for-build:

:exclude-for-build:gradle

To run the application, execute `./mvnw mn:run`.

:exclude-for-build:

[source,bash]
----
...
14:31:13.104 [main] INFO  io.micronaut.runtime.Micronaut - Startup completed in 506ms. Server Running: http://localhost:8082
----

Run `bookrecommendation` microservice:

:exclude-for-build:maven

To run the application, execute `./gradlew run`.

:exclude-for-build:

:exclude-for-build:gradle

To run the application, execute `./mvnw mn:run`.

:exclude-for-build:

[source,bash]
----
...
14:31:57.389 [main] INFO  io.micronaut.runtime.Micronaut - Startup completed in 523ms. Server Running: http://localhost:8080
----

You can run a cURL command to test the whole application:

[source, bash]
----
$ curl http://localhost:8080/books
[{"name":"Building Microservices"}
----

> You can then navigate to AWS Console and access the X-Ray UI

Reviewer: ```suggestion
You can then navigate to AWS Console and access the X-Ray UI.
```

The previous request generates such a trace:

image::xraytrace-map.png[]
image::xraytrace.png[]

In the previous image, you can see that:

- Whenever a Micronaut HTTP client executes a new network request, it creates a new subsegment.
- Whenever a Micronaut server receives a request, it creates a new segment.

Moreover, you can see the requests to `bookinventory` are made in parallel.

== Next steps

> As you have seen in this guide, you get distributing tracing up-and-running fast with the Micronaut framework without any annotations.

Reviewer: ```suggestion
As you have seen in this guide, you get distributed tracing up and running fast with the Micronaut framework without any annotations.
```
